diff --git a/app/vmauth/auth_config.go b/app/vmauth/auth_config.go
index b819152bd..3774691cd 100644
--- a/app/vmauth/auth_config.go
+++ b/app/vmauth/auth_config.go
@@ -440,11 +440,16 @@ func (bu *backendURL) setBroken() {
 }
 
 func (bu *backendURL) runHealthCheck() {
-	port := bu.url.Port()
-	if port == "" {
-		port = "80"
+	var network, addr string
+	if bu.url.Scheme == "unix" {
+		network, addr = "unix", strings.TrimPrefix(bu.url.String(), "unix:")
+	} else {
+		port := bu.url.Port()
+		if port == "" {
+			port = "80"
+		}
+		network, addr = "tcp", net.JoinHostPort(bu.url.Hostname(), port)
 	}
-	addr := net.JoinHostPort(bu.url.Hostname(), port)
 
 	t := time.NewTicker(*failTimeout)
 	defer t.Stop()
@@ -455,13 +460,13 @@ func (bu *backendURL) runHealthCheck() {
 			// Verify network connectivity via TCP dial before marking backend healthy.
 			// See https://github.com/VictoriaMetrics/VictoriaMetrics/issues/9997
 			ctx, cancel := context.WithTimeout(bu.bhc.ctx, time.Second)
-			c, err := netutil.Dialer.DialContext(ctx, "tcp", addr)
+			c, err := netutil.Dialer.DialContext(ctx, network, addr)
 			cancel()
 			if err != nil {
 				if errors.Is(bu.bhc.ctx.Err(), context.Canceled) {
 					return
 				}
-				logger.Warnf("ignoring the backend at %s for %s because of dial error: %s", addr, *failTimeout, err)
+				logger.Warnf("ignoring the backend at %s for %s because of dial error: %s:%s", network, addr, *failTimeout, err)
 				continue
 			}
 
@@ -1298,10 +1303,11 @@ func (up *URLPrefix) sanitizeAndInitialize() error {
 
 func sanitizeURLPrefix(urlPrefix *url.URL) (*url.URL, error) {
 	// Validate urlPrefix
-	if urlPrefix.Scheme != "http" && urlPrefix.Scheme != "https" {
+	if urlPrefix.Scheme != "http" && urlPrefix.Scheme != "https" && urlPrefix.Scheme != "unix" {
 		return nil, fmt.Errorf("unsupported scheme for `url_prefix: %q`: %q; must be `http` or `https`", urlPrefix, urlPrefix.Scheme)
 	}
-	if urlPrefix.Host == "" {
+	if urlPrefix.Host == "" &&
+		(urlPrefix.Scheme != "unix" || urlPrefix.Path == "") { // url.Parse fills only Path for unix:/a/b
 		return nil, fmt.Errorf("missing hostname in `url_prefix %q`", urlPrefix.Host)
 	}
 	return urlPrefix, nil
diff --git a/app/vmauth/auth_config_test.go b/app/vmauth/auth_config_test.go
index e0cd635a4..792d04f81 100644
--- a/app/vmauth/auth_config_test.go
+++ b/app/vmauth/auth_config_test.go
@@ -407,6 +407,8 @@ users:
   url_prefix:
   - http://node1:343/bbb
   - http://srv+node2:343/bbb
+  - unix:///absolute/path.socket
+  - unix:/short/path.socket
   tls_insecure_skip_verify: false
   retry_status_codes: [500, 501]
   load_balancing_policy: first_available
@@ -420,6 +422,8 @@ users:
 			URLPrefix: mustParseURLs([]string{
 				"http://node1:343/bbb",
 				"http://srv+node2:343/bbb",
+				"unix:///absolute/path.socket",
+				"unix:/short/path.socket",
 			}),
 			TLSInsecureSkipVerify:  &insecureSkipVerifyFalse,
 			RetryStatusCodes:       []int{500, 501},
@@ -836,7 +840,7 @@ func TestGetLeastLoadedBackendURL(t *testing.T) {
 	up := mustParseURLs([]string{
 		"http://node1:343",
 		"http://node2:343",
-		"http://node3:343",
+		"unix://node3/343.socket",
 	})
 	up.loadBalancingPolicy = "least_loaded"
 
@@ -909,7 +913,7 @@ func TestBrokenBackend(t *testing.T) {
 	up := mustParseURLs([]string{
 		"http://node1:343",
 		"http://node2:343",
-		"http://node3:343",
+		"unix://node3/343.socket",
 	})
 	up.loadBalancingPolicy = "least_loaded"
 	pbus := up.bus.Load()
diff --git a/lib/httpserver/httpserver.go b/lib/httpserver/httpserver.go
index 19784cb3a..06ad80195 100644
--- a/lib/httpserver/httpserver.go
+++ b/lib/httpserver/httpserver.go
@@ -140,7 +140,22 @@ func serve(addr string, rh RequestHandler, idx int, opts ServeOptions) {
 		}
 		tlsConfig = tc
 	}
-	ln, err := netutil.NewTCPListener(scheme, addr, useProxyProto, tlsConfig)
+	var ln net.Listener
+	var err error
+	if addr, ok := strings.CutPrefix(addr, "unix:"); ok {
+		if tlsConfig != nil {
+			logger.Fatalf("listening on unix domain socket (unix:%s) with TLS is not supported", addr)
+		}
+		ln, err = net.ListenUnix("unix", &net.UnixAddr{Net: "unix", Name: addr})
+		if err == nil {
+			chmodErr := os.Chmod(addr, 0770)
+			if chmodErr != nil {
+				logger.Infof("chmod %s failed: %s", addr, chmodErr)
+			}
+		}
+	} else {
+		ln, err = netutil.NewTCPListener(scheme, addr, useProxyProto, tlsConfig)
+	}
 	if err != nil {
 		logger.Fatalf("cannot start http server at %s: %s", addr, err)
 	}
diff --git a/lib/httpserver/httpserver_test.go b/lib/httpserver/httpserver_test.go
index 474ea4466..45f674135 100644
--- a/lib/httpserver/httpserver_test.go
+++ b/lib/httpserver/httpserver_test.go
@@ -1,11 +1,15 @@
 package httpserver
 
 import (
+	"context"
 	"encoding/json"
 	"net/http"
 	"net/http/httptest"
+	"os"
+	"path/filepath"
 	"strings"
 	"testing"
+	"time"
 
 	"github.com/VictoriaMetrics/VictoriaMetrics/lib/flagutil"
 )
@@ -229,3 +233,51 @@ func TestHandlerWrapper(t *testing.T) {
 		t.Fatalf("unexpected CSP header; got %q; want %q", got, cspHeader)
 	}
 }
+
+func TestServe(t *testing.T) {
+	hndl := RequestHandler(func(w http.ResponseWriter, r *http.Request) bool {
+		return true
+	})
+	addrs := make([]string, 0, 2)
+	ts := httptest.NewServer(http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) { hndl(w, r) }))
+	ts.Close()
+	addrs = append(addrs, ts.URL)
+	dn, err := os.MkdirTemp("", "httpserver-test-*")
+	if err != nil {
+		t.Fatal(err)
+	}
+	defer os.RemoveAll(dn)
+	addrs = append(addrs, "unix:"+filepath.Join(dn, "socket"))
+	tr := http.DefaultTransport.(*http.Transport).Clone()
+	unixtransportRegister(tr)
+	unixClient := &http.Client{Transport: tr}
+
+	// t.Log("addrs:", addrs)
+	for _, addr := range addrs {
+		scheme, _, _ := strings.Cut(addr, ":")
+		t.Run(scheme, func(t *testing.T) {
+			serveAddr := strings.TrimPrefix(addr, "http://")
+			go func() {
+				serve(serveAddr, hndl, 0, ServeOptions{})
+				t.Log("returned")
+			}()
+			defer stop(serveAddr)
+			time.Sleep(100 * time.Millisecond)
+			client := http.DefaultClient
+			if scheme == "unix" {
+				client, addr = unixClient, "http+"+addr
+			}
+			ctx, cancel := context.WithTimeout(t.Context(), time.Second)
+			defer cancel()
+			req, err := http.NewRequestWithContext(ctx, "GET", addr, nil)
+			if err != nil {
+				t.Fatal(err)
+			}
+			resp, err := client.Do(req)
+			if err != nil {
+				t.Fatalf("%s %s (%s): %+v", req.Method, req.URL, addr, err)
+			}
+			t.Log(resp.Status)
+		})
+	}
+}
diff --git a/lib/httpserver/httpserver_unixtransport_test.go b/lib/httpserver/httpserver_unixtransport_test.go
new file mode 100644
index 000000000..7c7c03c8c
--- /dev/null
+++ b/lib/httpserver/httpserver_unixtransport_test.go
@@ -0,0 +1,98 @@
+// Copied from https://github.com/peterbourgon/unixtransport/blob/v0.0.7/register.go
+//
+// SPDX-License-Identifier: Apache-2.0
+
+package httpserver
+
+import (
+	"context"
+	"encoding/base64"
+	"fmt"
+	"net"
+	"net/http"
+	"strings"
+)
+
+// Register adds a protocol handler to the provided transport that can serve
+// requests to Unix domain sockets via the "http+unix" or "https+unix" schemes.
+// Request URLs should have the following form:
+//
+//	https+unix:///path/to/socket:/request/path?query=val&...
+//
+// The registered transport is based on a clone of the provided transport, and
+// so uses the same configuration: timeouts, TLS settings, and so on. Connection
+// pooling should also work as normal. One caveat: only the DialContext and
+// DialTLSContext dialers are respected; the Dial and DialTLS dialers are
+// explicitly removed and ignored. Any configured Proxy is also discarded.
+func unixtransportRegister(t *http.Transport) {
+	copy := t.Clone()
+
+	copy.Dial = nil    //lint:ignore SA1019 yes, it's deprecated, that's the point
+	copy.DialTLS = nil //lint:ignore SA1019 yes, it's deprecated, that's the point
+	copy.DialTLSContext = nil
+	copy.Proxy = nil // Proxy doesn't support Unix sockets, so drop it
+
+	switch {
+	case copy.DialContext == nil:
+		copy.DialContext = dialContextAdapter((&net.Dialer{}).DialContext)
+
+	case copy.DialContext != nil:
+		copy.DialContext = dialContextAdapter(copy.DialContext)
+
+	}
+
+	tt := roundTripAdapter(copy)
+
+	t.RegisterProtocol("http+unix", tt)
+}
+
+// dialContextAdapter decorates the provided DialContext function by trying to base64 decode
+// the provided address. If successful, the network is changed to "unix" and the address
+// is changed to the decoded value.
+func dialContextAdapter(next dialContextFunc) dialContextFunc {
+	return func(ctx context.Context, network, address string) (net.Conn, error) {
+		host, _, err := net.SplitHostPort(address)
+		if err != nil {
+			host = address
+		}
+
+		filepath, err := base64.RawURLEncoding.DecodeString(host)
+		if err == nil {
+			network, address = "unix", string(filepath)
+		}
+
+		return next(ctx, network, address)
+	}
+}
+
+// roundTripAdapter returns an http.RoundTripper which, when used in combination
+// with the dialContextAdapter, supports Unix sockets via any scheme with a
+// "+unix" suffix.
+func roundTripAdapter(next http.RoundTripper) http.RoundTripper {
+	return roundTripFunc(func(req *http.Request) (*http.Response, error) {
+		if req.URL == nil {
+			return nil, fmt.Errorf("unix transport: no request URL")
+		}
+
+		scheme := strings.TrimSuffix(req.URL.Scheme, "+unix")
+		if scheme == req.URL.Scheme {
+			return nil, fmt.Errorf("unix transport: missing '+unix' suffix in scheme %s", req.URL.Scheme)
+		}
+
+		socketPath, requestPath, _ := strings.Cut(req.URL.Path, ":")
+		encodedHost := base64.RawURLEncoding.EncodeToString([]byte(socketPath))
+
+		req = req.Clone(req.Context())
+		req.URL.Scheme = scheme
+		req.URL.Host = encodedHost
+		req.URL.Path = requestPath
+
+		return next.RoundTrip(req)
+	})
+}
+
+type dialContextFunc func(ctx context.Context, network, address string) (net.Conn, error)
+
+type roundTripFunc func(req *http.Request) (*http.Response, error)
+
+func (f roundTripFunc) RoundTrip(req *http.Request) (*http.Response, error) { return f(req) }
